### PR TITLE
sysoptions: search XDG_DATA_HOME and XDG_DATA_DIRS for vulkan ICDs

### PR DESCRIPTION
**Pull request by LunNova**
* Search XDG_*_DIRS for vulkan ICDs
  * This is needed to match Vulkan-Loader's behavior, otherwise on some platforms ICDs are not found.
  * See https://github.com/KhronosGroup/Vulkan-Loader/blob/v1.3.235/docs/LoaderDriverInterface.md#driver-discovery-on-linux for details of the vulkan loader search paths.
* Use vulkaninfo when available to find device info and only use glxinfo as a fallback.
  * This ensures the device printed for a vulkan ICD is actually the device it'll use.
* Ensure default/empty choice does not set a vulkan ICD override instead of selecting the first one.

I wanted to make it clear which GPU is associated with a given ICD where possible.

This approach does make the picker a bit cluttered for multi-gpu setups. If prime has an impact for a particular ICD both the default and DRI_PRIME GPU are shown which can result in a very long string.

image

### CODE REVIEW

**Comment by LunNova** [comment]
Marked as draft because not sure what to do about very long vendor strings for multi-gpu without losing info in the output.

File: lutris/sysoptions.py
Comment (LunNova) [inline lutris/sysoptions.py:131]: This removes the VK_ICD_FILENAMES logic as the new logic is able to show what Auto will do if this is already set without any special casing - vulkaninfo will run with it set and show which GPU will be used.

It also allows non-auto options to override it even if it is set system-wide. If that's not desired I can add a check back and have all values default to "" if it's set.

In that case I'd also add `VK_DRIVER_FILES` and make `VK_ICD_FILENAMES` the fallback per the vulkan loader spec.

File: lutris/sysoptions.py
Comment (LunNova) [inline lutris/sysoptions.py:187]: This is gone as it'll now show `Auto: (your default GPU)` so it actually tells you what the default is.

**Comment by LunNova** [comment]
This approach showing the GPU next to the ICD probably breaks translations too :/

**Comment by LunNova** [comment]
This is marked draft as it needs some help/input, but is ready for review. Github doesn't distinguish between "don't merge" and "don't review" draft states.

File: lutris/sysoptions.py
Comment (danieljohnson2) [inline lutris/sysoptions.py:114]: This is not how Lutris usually handles environments.

'env' would normally be a dict, and passed as the 'env' parameter of Popen, and os.environ.copy() to obtain the parent enviroment first.

If we must prefix some shell stuff - I don't really like it - then at least the parameter name should not be 'env'. It is asking for a dict to be passed as per convention.

File: lutris/sysoptions.py
Comment (danieljohnson2) [reply lutris/sysoptions.py:131]: If vulkaninfo is not available, then I think VK_ICD_FILENAMES will again be ignored, and issue #4610 will be back.

So, how common is it vulkaninfo?

File: lutris/sysoptions.py
Comment (danieljohnson2) [inline lutris/sysoptions.py:351]: No default? That seems unfortunate.

File: lutris/sysoptions.py
Comment (danieljohnson2) [inline lutris/sysoptions.py:138]: This logic does not entirely agree with [your link](https://github.com/KhronosGroup/Vulkan-Loader/blob/v1.3.235/docs/LoaderDriverInterface.md#driver-discovery-on-linux). That document describes how it falls back to other things when those environment variables are not available.

I think your logic falls back even if they are, I think.

File: lutris/sysoptions.py
Comment (danieljohnson2) [inline lutris/sysoptions.py:152]: I don't like this sorted(); I think the link described a search through the directories in a well defined order and that was *not* alphabetical by path. I think alphabetical by path is very surprising and not desirable.

I think you should be using ordered sets throughout.

**Comment by danieljohnson2** [comment]
I think @strycore would prefer to look at a finished product, that is ready to merge.

I don't think is, though. I found several issues that I'd like to see improvement on before I'd take this merge.

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py:351]: Defaulting to "" makes the default Auto which leaves the env var unset.

IMO this makes more sense than Lutris defaulting to setting this env var, which is the current behavior.

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py:138]: That's a good point, I'll try to make the search order match that more exactly.

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py:131]: That particular issue wouldn't recur as Auto now leaves the env var unset, instead of trying to set it to whatever the default already was.

However if vulkaninfo isn't available and the system has multiple GPUs the text in the list showing which GPU corresponds with which ICD may be wrong.

Would it make more sense to not show which specific GPU corresponds to which ICD if vulkaninfo isn't available?

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py:152]: The order here only impacts the order of multiple different options for a single vendor showing up in the dropdown.

It has no impact on the default selection since we default to leaving the env var unset.

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py:351]: I could revert this line change - get_vk_icd_choices()[0][1] -> ("Auto: the GPU name", "")[1] -> "" so it makes no difference other than being slightly faster.

Or maybe add a comment noting that this corresponds to Auto / not setting the env var to allow default device selection to happen.

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py:138]: Hmm, not sure the ordering matters here: the order matters for the vulkan loader selecting a default device.
It doesn't matter here because we just want to show all available ICDs for the user to pick between.

What about adding the extra fallback options of $HOME/.config and /etc/xdg and /etc and /usr/share that are currently missing but not trying to do anything about the ordering?

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py:152]: Oh right, it can impact order for multiple ones for a single vendor so I have to fix that.

Will get back to you with that.

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py:114]: Got it, I'll sort this out.

File: lutris/sysoptions.py
Comment (danieljohnson2) [reply lutris/sysoptions.py:351]: Oh, I see. Perhaps a comment then, I missed that.

File: lutris/sysoptions.py
Comment (danieljohnson2) [reply lutris/sysoptions.py:131]: It would be nice to know what is being selected, if that's possible.

But if vulkaninfo is universal anyway it may not really matter.

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py]: Should be fixed now.

File: lutris/sysoptions.py
Comment (LunNova) [reply lutris/sysoptions.py]: Swapped it to use the env parameter.

**Comment by LunNova** [comment]
Tried to address feedback by following the spec much more precisely.

Shortened the text in the dropdown by removing the dev/chip name part from the vulkaninfo output.

image

File: lutris/sysoptions.py
Comment (danieljohnson2) [inline lutris/sysoptions.py:143]: Perhaps os.path.join would be better here?

**Review by danieljohnson2 — APPROVED** [review]
I like this better. It's not the same behavior as before, and that may cause compatibility issues. But on the whole I think this looks like an improvement.

Still, @strycore will have to make the final call on this.

**Comment by strycore** [comment]
I don't really have a way to test this so it may be better if @GloriousEggroll reviews this one as he's written most of the code for Vulkan loaders.

**Pull request merged**
